### PR TITLE
tests: add networkd integration test

### PR DESCRIPTION
**Pull request by martinpitt**
This uses temporary configuration in /run and temporary veth devices, and does not write anything on disk or change any system configuration; but it assumes (and checks at the beginning) that networkd is not currently running.

This can be run on a normal installation, in QEMU, nspawn, or LXC. As this requires root privileges, this is not integrated into "make check".

This comes from Debian's/Ubuntu's autopkgtest, but this particular one is not really distro specific. It has been helpful to find a number of regressions (at least issues #1147, #1645, #1866, and a brand new regression with `DHCP=yes` which isn't filed as an issue yet), and @teg is now using these tests as well (in Fedora nspawn), so I figure it's time to put them into the upstream tree.

All of these are passing on 227 and are very robust there. Some of them fail on current master, which are actual regressions that are being worked on.

### CODE REVIEW

**Comment by martinpitt** [comment]
Possible future improvements:
- Optionally run against the built source tree; either we start networkd directly instead of through `systemctl start`, or more elegantly, copy the system unit to `/run` and modify the path to the binary in it.
- Integrate into the upstream QEMU tests. This doesn't work very well for me, so maybe someone else wants to pick this up? It essentially needs `iproute2`, `python3`, optionally `dnsmasq`, and the test script, and of course networkd.

**Comment by poettering** [comment]
Urks, dnsmasq... Any chance this can be reworked to test this against networkd's own dhcp server? That way we could test this much more automatically...

**Comment by martinpitt** [comment]
I'd advise against testing networkd against networkd's (I suppose you actually mean resolved?) DHCP server. The systemd components share way too much code in the private libsystemd, so a protocol implementation error could affect both ends in the same way and thereby hiding it. This is why I prefer testing against an independent DHCP server. dnsmasq is small, widespread, does not need any configuration file, and simple to set up in a test.

**Comment by poettering** [comment]
Well, sure, it's more expressive to test against other DHCP server, but it effectively means it won't be tested as much, because it requires external dependencies. If something requires installing of an external daemon like this on my machine I won't run the test for example.

**Comment by poettering** [comment]
And no, the DHCP server is in networkd itself. See the DHCPServer=yes option. resolved is a DNS client, that's all.

**Comment by martinpitt** [comment]
> And no, the DHCP server is in networkd itself.

Ah, ok. That would mean to test networkd against itself. That sounds like a really bad idea, and it's also not really a realistic scenario.

> If something requires installing of an external daemon like this on my machine

You'd usually run this in nspawn or QEMU, but yes -- having some external dependencies is the price for that.

What perhaps could work is to start a second networkd process (not `systemd-networkd.service`), configure that appropriately, run it in its own mount namespace and hide the main system's `/{etc,lib,run}/systemd/network/` from it, so that it by and large acts like `dnsmasq` in the current test. That would avoid to test the "main" networkd process against itself and be a realistic scenario. But it still doesn't solve the problem that you can possibly hide protocol implementation errors.

Bottom line: sorry, TBH I'm not very interested in doing this. It makes the setup much more complicated and fragile and the test less useful.

**Comment by poettering** [comment]
> Ah, ok. That would mean to test networkd against itself. That sounds like a really bad idea, and it's also not really a realistic scenario.

Not realistic? It's pretty realistic. In fact, it's the default out-of-the-box when networkd is used in conjunction with nspawn -n. systemd upstream comes with a pair of .network files that apply on the host and in the container, where the host does DHCPServer=yes, and the container does DHCP=yes.

Why would this be a "bad" idea? I mean, again, testing against foreign software of course is more expressive, but what's that good for if we never run that?

**Comment by martinpitt** [comment]
> systemd upstream comes with a pair of .network files that apply on the host and in the container, where the host does DHCPServer=yes, and the container does DHCP=yes.

That means that the DHCP server is on the host, right? I. e. the DHCP-serving `networkd` is an entirely different process than the DHCP client in the nspawn container. That's what I suggested in the "start a second networkd process ... in its own mount namespace" above, as for the test we can't assume to have (or control) an "outside" host -- everything needs to happen within the test environment. It is IMHO really not an realistic test to have the _same_ netword process in the test environment be the  DHCP server and client.

> what's that good for if we never run that

Tom doesn't seem to have a problem with it :-) Changing it for this would mean that we have to maintain two tests, one for downstream against a separate implementation (which I really insist on keeping at least for the downstream tests), and one for upstream against networkd. Doable, but please understand that I'm not particuarly thrilled by that.

So, let's be honest: I'm perfectly happy if you say "I don't like the test in this form because of the external dependency" and don't want to merge it. If so, please close this and let's move on.

**Comment by martinpitt** [comment]
OK, to turn that into something constructive: I could live with a logic like

```
if (dnsmasq installed)
    set up dnsmasq DHCP server
else
    set up new mount ns with separate s-networkd as DHCP server
```

I won't personally test the second path a lot, but if you want this, then I can work on this.

**Comment by mbiebl** [comment]
@martinpitt couldn't we test both? Somehow it sounds this might be useful to test against an external _and_ internal implementation. And the test against dnsmasq could be conditional.

So everything I would change from your proposal would be to s/else/fi/

**Comment by poettering** [comment]
well, if @teg likes this, that's good enough. I personally think a test case that uses networkd to set up a veth link and then does dhcp server on one end of it, and dhcp client on the other would be a great test with a good chance that people (specifically: me personally) will run frequently, as it is self-contained. Also, it tests both DHCP server and client. Of course, it would ultimately not have the same expressivity, as it test our code against our code, but it tests different code paths as a test that just runs dnsmasq, and certainly has a bigger chance to actually be run...

(the test would work with a single networkd instance, where the dhcp server and client parts wouldn't know that they actually are runing in the same daemon.

**Comment by martinpitt** [comment]
Ack, I'll look into that.

**Comment by teg** [comment]
I must say I like running the test against dnsmasq (for the reasons @martinpitt mentioned), though I guess making this a soft dep would make sense. Would be even better to additionally run the test against ourselves, but I would make this an addition rather than a fallback.

**Comment by martinpitt** [comment]
I reworked the tests as above:
- There are now two test classes for dnsmasq and networkd (in "router config" mode), and the latter is isolated from the main `systemd-networkd.service` (which we use for the "client" side) by running in a separate mount namespace and hiding the main system's network config and system D-Bus.
- dnsmasq tests get cleanly skipped if dnsmasq is not available.
- Note that networkd does not currently support DHCPv6 server, and I also did not find a way to hand the "client" an IPv6 address via RA; so the two IPv6 tests get skipped with networkd and only work with dnsmasq. @teg, did I miss anything here?
- If a test fails, the debug logs now only include the journal entries for the current test case, not the whole journal since the boot of the machine (using the journal cursor). This makes reading the logs much easier.

**Comment by martinpitt** [comment]
Force-pushed again with a race condition fix with the IPv6 tests. s-n-wait-online already exits after it gets an IPv6 _or_ IPv4 address, but we assert both, so we need to manually wait for the other.

**Comment by poettering** [comment]
Looks good to me. But I'll leave this one to @teg. @teg, say something or just merge it.

**Comment by keszybz** [comment]
LGTM.

**Comment by teg** [comment]
Sorry for the delay, got lost. Looks good to me. Oh, and @martinpitt, no you did not miss anything, we do not support any server stuff for IPv6.
